### PR TITLE
Type annotations round 2

### PR DESCRIPTION
**Pull request by bmerry**
Sprinkles more type annotations around. There is one major functional change: `Operation.buffer` now raises an exception if there is no buffer bound, rather than returning None. It's pretty much always called with a static string for which a buffer is known to be present, so I don't expect this to cause much if any breakage, and it allows the return type to be `DeviceArray` instead of `Optional[DeviceArray]`.

### CODE REVIEW

File: katsdpsigproc/fill.py [hunk 1]
```diff
@@ -43,20 +47,21 @@ def __init__(self, context, dtype, ctype, tuning=None):
...
             fn = cls(context, dtype, ctype, {'wgs': wgs}).instantiate(queue, shape)
             fn.bind(data=data)
             return tune.make_measure(queue, fn)
 
         return tune.autotune(generate, wgs=[64, 128, 256, 512])
 
-    def instantiate(self, *args, **kwargs):
-        return Fill(self, *args, **kwargs)
+    def instantiate(self, command_queue: AbstractCommandQueue, shape: Tuple[int, ...],
+                    allocator: Optional[accel.AbstractAllocator] = None) -> 'Fill':
```
Comment (ludwigschwardt) [inline katsdpsigproc/fill.py]: Does `'Fill'` get linked to the `Fill` class by mypy?

File: katsdpsigproc/percentile.py [hunk 2]
```diff
@@ -72,17 +83,23 @@ def generate(size, wgsy):
             inp.set(queue, host_data)
             return tune.make_measure(queue, fn)
 
-        return tune.autotune(generate,
-                             size=[8, 16, 32, 64, 128, 256, 512, 1024],
-                             wgsy=[1, 2, 4, 8, 16, 32])
+        return cast(_TuningDict, tune.autotune(generate,
```
Comment (ludwigschwardt) [inline katsdpsigproc/percentile.py]: Didn't think I'd see `cast` in Python, but hey :-)

File: katsdpsigproc/test/test_accel.py [hunk 3]
```diff
@@ -331,53 +343,55 @@ def test_raw(self, context, queue):
...
+    def test_coherence(self, context: AbstractContext, queue: AbstractCommandQueue) -> None:
+        """Check that runtime correct transfers data between the host and device views."""
         source = """\
             <%include file="/port.mako"/>
             KERNEL void triple(unsigned int *data)
             {
                 data[get_global_id(0)] *= 3;
             }"""
-        program = accel.build(context, None, source=source)
+        program = accel.build(context, 'not a file', source=source)
```
Comment (ludwigschwardt) [inline katsdpsigproc/test/test_accel.py]: Why this change?

File: katsdpsigproc/fill.py (hunk 1 above)
Comment (bmerry) [reply katsdpsigproc/fill.py]: Yes.

File: katsdpsigproc/percentile.py (hunk 2 above)
Comment (bmerry) [reply katsdpsigproc/percentile.py]: Yep :-) It has no runtime effect though - it's just a hint to mypy.

File: katsdpsigproc/test/test_accel.py (hunk 3 above)
Comment (bmerry) [reply katsdpsigproc/test/test_accel.py]: mypy was complaining because the parameter is annotated as `str` and `None` isn't a `str`.
